### ISSUE TITLE
ART classifier tanks Pytorch Model Accuracy

### ISSUE THREAD

**Issue by MaxLikesMath** [Trusted-AI/adversarial-robustness-toolbox#895]
**Describe the bug**
When passing PyTorch model to ART classifier, the new classifier class tanks the model accuracy for some reason.

**To Reproduce**
1. Train a Pytorch model, have the forward pass return torch.nn.functional.log_softmax(x) where x is the output of the last layer.
2. Pass model to the ART Pytorch classifier class.
3. Load in test dataset as collection of np.arrays
4. Get model predictions
5. compute accuracy

**Expected behavior**
The accuracy of the original Pytorch model will be much better. The ART classifier downgrades accuracy down to near random.

**System information (please complete the following information):**
 - Windows
 - Python version 3.6. something
 - ART version 1.5
 - PyTorch 1.6

**Comment by beat-buesser** [comment]
Hi @MaxLikesMath Could you please share the code reproducing your observation?

**Comment by MaxLikesMath** [comment]
`test_preprocessing = transforms.Compose([
    transforms.ToTensor(),
])

batch_size=256

cifar10_test  = dsets.CIFAR10(root='./Datasets', train=False,
                              download=True, transform=test_preprocessing)

test_loader = torch.utils.data.DataLoader(cifar10_test,
                                          batch_size=batch_size,
                                          shuffle=False)

import cifar_models

num_classes=10
model_loc = '/root/AttackEngineTesting/Saved_Models/ResNet18_1_CIFAR10.pt'
model_ft = cifar_models.resnet.ResNet18(num_classes=num_classes)
device = multigpu_cuda(model_ft)
model_ft.load_state_dict(torch.load(model_loc)['state_dict'])
optimizer = optim.SGD(model_ft.parameters(), 0.001)

from art.estimators.classification import PyTorchClassifier

x,y = GetBatchesAsNumpy(test_loader, num_batches=15)

'''
When declaring classifier in deployment, pass dataset stats as preprocessing args
'''
art_classifier = PyTorchClassifier(
    model=model_ft,
    loss=nn.NLLLoss(),
    clip_values=(0, 1),
    optimizer=optimizer,
    input_shape=(3, 32, 32),
    nb_classes=10,
)

preds = art_classifier.predict(x)

class_preds = np.argmax(preds, axis=1)

acc = sklearn.metrics.accuracy_score(y, class_preds)

`
Thanks for the quick response. The above code is what runs the ART classifier. It seems to decrease the test accuracy from ~80% when using native Pytorch to about 10% with the ART classifier class. I have reproduced it with 2 models on 2 different datasets as well.

**Comment by beat-buesser** [comment]
Great, thank you. I haven't run it yet because of the model checkpoint. Are you running the evaluation of the PyTorch model achieving 80% with the same script and the same way of loading the data?

**Comment by MaxLikesMath** [comment]
I have two scripts, one which trains the model, and one which loads it, turns it into the art classifier, and tests it as an art classifier. It is preprocessed the same in both scripts. The only difference in how the model is evaluated is that the data is converted to numpy arrays from Torch tensors in the ART script as is required by the ART classifier.

**Comment by beat-buesser** [comment]
Can you share `GetBatchesAsNumpy`? I would like to see if I can reproduce it.

**Comment by MaxLikesMath** [comment]
Yup! Sorry, I thought I had included it:

def GetBatchesAsNumpy(loader, num_batches=15):
    '''
    A script utility function for getting batches from the PyTorch dataloader as numpy arrays. Won't be needed in production.
    '''
    test_dataset_array = []
    test_dataset_labels = []
    data_iter = iter(loader)
    
    for batch in range(num_batches):
        try:
            data_batch = next(data_iter)[0].numpy()
            label_batch = next(data_iter)[1].numpy()
            test_dataset_array.append(data_batch)
            test_dataset_labels.append(label_batch)
        except:
            break

    test_dataset_array = np.concatenate(test_dataset_array, axis=0)
    test_dataset_labels = np.concatenate(test_dataset_labels, axis=0)
    return test_dataset_array, test_dataset_labels

My intuition is that somewhere the data is getting altered in some way, but I am not sure.

**Comment by beat-buesser** [comment]
Thank you. My first check would be to see the range and shape of `x`. I have seen previously cases where the transformations by `transforms.ToTensor()` led to changes not expected by the user.

**Comment by MaxLikesMath** [comment]
The shape is normal and is (num_samples, 3, 32,32), and the range is [0,1]. I know transforms.ToTensor will always scale a PIL image from [0,1] as well. When the classifier is declared in ART, it takes in certain preprocessing arguments. Is there any functionality when data is passed into the predict function that could be causing the data to undergo some unexpected input transformation?

**Comment by beat-buesser** [comment]
This one took me a moment, but I think I found the source of the issue: `GetBatchesAsNumpy` calls next() twice on `data_iter` to get `data_batch` and `label_batch`. this means that these two variables do not belong to the same sample/label pair, but are shifted b by 1 sample. That's why the performance becomes completely random. This solved it for me
```
data = next(data_iter)
data_batch = data[0].numpy()
label_batch = data[1].numpy()
```

**Comment by MaxLikesMath** [comment]
Well, that would do it. Silly error on my part. Thank you!
